src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -68,80 +77,286 @@ import           Database.LSMTree.Internal.Serialise
+
+  Let's compare the memory size of @Index2@ with @Index3@. Say we have @O(n)@
+  pages, and keys that are @k@ bits each, then the memory size of @Index2@ is
+  @O(nk)@ bits. Say we have @O(n)@ pages and we store only the @m@ most
+  significant bits, then the memory size of @Index3@ is @O(n m
+  + 2 n + E[collision n m] * k)@ bits, where the last summand is the expected
+  number of collisions for @n@ independently generated hashes of bit-size @m@.
+  So, @Index3@ is not strictly an improvement over @Index2@ if we look at memory
+  complexity alone. However, /in practice/ @Index3@ is an improvement over
+  @Index2@ if we can pick an @m@ that is (i) much smaller than @k@ and (ii)
+  keeps @E[collision n m]@ small (if not close to @0@).
+
+  For example, storing the first @48@ bits of @2.5@ million SHA256 hashes
+  reduces memory size from @n * 256@ bits to @n * 48@ bits, because the expected
+  number of collisions is smaller than @1@.
+
+  TODO: the expected number of collisions is (n^2 - n ) / (2 * m) for hashing
+  @n@ distinct values into @m@ distinct buckets. Is the following reasoning
+  correct?
+
+  \[
+  \begin{align*}
+     &~ O \left( \frac{n^2 - n}{2m}                                                 \right) \\
+    =&~ O \left( \left( n^n - n \right) *   \left( \frac{1}{2m}             \right) \right) \\
+    =&~ O        \left( n^n - n \right) * O \left( \frac{1}{2m}                     \right) \\
+    =&~ O        \left( n^2 - n \right) * O \left( \left( 2m \right)^{ -1 }         \right) \\
+    =&~ O        \left( n^2 - n \right) * O \left( 2^{ -1} * m^{ -1}                \right) \\
+    =&~ O        \left( n^2 - n \right) * O \left( m^{ -1}                          \right) \\
+    =&~ O        \left( n^2     \right) * O \left( m^{ -1}                          \right) \\
+    =&~ O \left(        n^2             *          m^{ -1}                          \right) \\
+    =&~ O \left( \frac{n^2}{m}                                                      \right) \\
+  \end{align*}
+  \]
+
+  ==== Range-finder
+
+  TODO: explain range-finder bits
+
+  === A semi-formal description of the compact index #rep-descr#
+
+  * \(n\) is the number of pages
+  * \(ps = \{p_i \mid 0 \leq i < n \}\) is a sorted set of pages
+  * \(p^{min}_i\) is the full minimum key on a page \(p_i \in ps\).
+  * \(p^{max}_i\) is the full maximum key on a page \(p_i \in ps\).
+  * \(m\) is the range-finder bit-precision
+  * \(\texttt{topBits16}(m, k)\) extracts the @m@ most significant bits from
+    @k@. We call these @m@ bits the range-finder bits.
+  * \(\texttt{sliceBits32}(m, k)\) extracts the @32@ most significant bits from @k@
+    /after/ @m@. We call these @32@ bits the primary bits.
+  * \(i \in \left[0, n \right)\), unless stated otherwise
+  * \(j \in \left[0, 2^m\right)\), unless stated otherwise
+  * Pages must be partitioned: \(\forall p \in ps. \texttt{topBits16}(m, p^{min}_x) ~\texttt{==}~ \texttt{topits16}(m,p^{max}_x) \)
```

Comment:
```suggestion
  * Pages must be partitioned: \(\forall p_x \in ps. \texttt{topBits16}(m, p^{min}_x) ~\texttt{==}~ \texttt{topits16}(m,p^{max}_x) \)
```
or alternatively drop the `_x` subscript on the RHS
```suggestion
  * Pages must be partitioned: \(\forall p \in ps. \texttt{topBits16}(m, p^{min}) ~\texttt{==}~ \texttt{topits16}(m,p^{max}) \)
```

Reply:
Changed it to `_i`